File: internal/plugins/jest/rules/no_identical_title/no_identical_title_test.go
package no_identical_title_test

import (
	"testing"

	"github.com/web-infra-dev/rslint/internal/plugins/jest/fixtures"
	"github.com/web-infra-dev/rslint/internal/plugins/jest/rules/no_identical_title"
	"github.com/web-infra-dev/rslint/internal/rule_tester"
)

func TestNoIdenticalTitleRule(t *testing.T) {
	rule_tester.RunRuleTester(
		fixtures.GetRootDir(),
		"tsconfig.json",
		t,
		&no_identical_title.NoIdenticalTitleRule,
		[]rule_tester.ValidTestCase{
			{Code: "it(); it();"},
			{Code: "describe(); describe();"},
			{Code: "describe(\"foo\", () => {}); it(\"foo\", () => {});"},
			{Code: `describe("foo", () => {
  it("works", () => {});
});
`},
			{Code: `it('one', () => {});
it('two', () => {});
`},
			{Code: `describe('foo', () => {});
describe('foe', () => {});
`},
			{Code: "it(`one`, () => {});\nit(`two`, () => {});\n"},
			{Code: "describe(`foo`, () => {});\ndescribe(`foe`, () => {});\n"},
			{Code: `describe('foo', () => {
  test('this', () => {});
  test('that', () => {});
});
`},
			{Code: `test.concurrent('this', () => {});
test.concurrent('that', () => {});
`},
			{Code: `test.concurrent('this', () => {});
test.only.concurrent('that', () => {});
`},
			{Code: `test.only.concurrent('this', () => {});
test.concurrent('that', () => {});
`},
			{Code: `test.only.concurrent('this', () => {});
test.only.concurrent('that', () => {});
`},
			{Code: `test.only('this', () => {});
test.only('that', () => {});
`},
			{Code: `describe('foo', () => {
  it('works', () => {});

  describe('foe', () => {
    it('works', () => {});
  });
});
`},
			{Code: `describe('foo', () => {
  describe('foe', () => {
    it('works', () => {});
  });

  it('works', () => {});
});
`},
			{Code: "describe('foo', () => describe('foe', () => {}));"},
			{Code: `describe('foo', () => {
  describe('foe', () => {});
});

describe('foe', () => {});
`},
			{Code: "test(\"number\" + n, function() {});"},
			{Code: "test(\"number\" + n, function() {}); test(\"number\" + n, function() {});"},
			{Code: "it(`${n}`, function() {});"},
			{Code: "it(`${n}`, function() {}); it(`${n}`, function() {});"},
			{Code: `describe('a class named ' + myClass.name, () => {
  describe('#myMethod', () => {});
});

describe('something else', () => {});
`},
			{Code: `describe('my class', () => {
  describe('#myMethod', () => {});
  describe('a class named ' + myClass.name, () => {});
});
`},
			{Code: "describe(\"foo\", () => {\n  it(`ignores ${someVar} with the same title`, () => {});\n  it(`ignores ${someVar} with the same title`, () => {});\n});\n"},
			{Code: "const test = { content: () => \"foo\" };\ntest.content(`something that is not from jest`, () => {});\ntest.content(`something that is not from jest`, () => {});\n"},
			{Code: "const describe = { content: () => \"foo\" };\ndescribe.content(`something that is not from jest`, () => {});\ndescribe.content(`something that is not from jest`, () => {});\n"},
			{Code: "describe.each`\n  description\n  ${'b'}\n`('$description', () => {});\n\ndescribe.each`\n  description\n  ${'a'}\n`('$description', () => {});\n"},
			{Code: "describe('top level', () => {\n  describe.each``('nested each', () => {\n    describe.each``('nested nested each', () => {});\n  });\n\n  describe('nested', () => {});\n});\n"},
			{Code: "describe.each``('my title', value => {});\ndescribe.each``('my title', value => {});\ndescribe.each([])('my title', value => {});\ndescribe.each([])('my title', value => {});\n"},
			{Code: `describe.each([])('when the value is %s', value => {});
describe.each([])('when the value is %s', value => {});
`},
		},
		[]rule_tester.InvalidTestCase{
			{
				Code: `describe('foo', () => {
  it('works', () => {});
  it('works', () => {});
});
`,
				Errors: []rule_tester.InvalidTestCaseError{
					{MessageId: "multipleTestTitle", Line: 3, Column: 6},
				},
			},
			{
				Code: `it('works', () => {});
it('works', () => {});
`,
				Errors: []rule_tester.InvalidTestCaseError{
					{MessageId: "multipleTestTitle", Line: 2, Column: 4},
				},
			},
			{
				Code: `test.only('this', () => {});
test('this', () => {});
`,
				Errors: []rule_tester.InvalidTestCaseError{
					{MessageId: "multipleTestTitle", Line: 2, Column: 6},
				},
			},
			{
				Code: `xtest('this', () => {});
test('this', () => {});
`,
				Errors: []rule_tester.InvalidTestCaseError{
					{MessageId: "multipleTestTitle", Line: 2, Column: 6},
				},
			},
			{
				Code: `test.only('this', () => {});
test.only('this', () => {});
`,
				Errors: []rule_tester.InvalidTestCaseError{
					{MessageId: "multipleTestTitle", Line: 2, Column: 11},
				},
			},
			{
				Code: `test.concurrent('this', () => {});
test.concurrent('this', () => {});
`,
				Errors: []rule_tester.InvalidTestCaseError{
					{MessageId: "multipleTestTitle", Line: 2, Column: 17},
				},
			},
			{
				Code: `test.only('this', () => {});
test.concurrent('this', () => {});
`,
				Errors: []rule_tester.InvalidTestCaseError{
					{MessageId: "multipleTestTitle", Line: 2, Column: 17},
				},
			},
			{
				Code: `describe('foo', () => {});
describe('foo', () => {});
`,
				Errors: []rule_tester.InvalidTestCaseError{
					{MessageId: "multipleDescribeTitle", Line: 2, Column: 10},
				},
			},
			{
				Code: `describe('foo', () => {});
xdescribe('foo', () => {});
`,
				Errors: []rule_tester.InvalidTestCaseError{
					{MessageId: "multipleDescribeTitle", Line: 2, Column: 11},
				},
			},
			{
				Code: `fdescribe('foo', () => {});
describe('foo', () => {});
`,
				Errors: []rule_tester.InvalidTestCaseError{
					{MessageId: "multipleDescribeTitle", Line: 2, Column: 10},
				},
			},
			{
				Code: `describe('foo', () => {
  describe('foe', () => {});
});
describe('foo', () => {});
`,
				Errors: []rule_tester.InvalidTestCaseError{
					{MessageId: "multipleDescribeTitle", Line: 4, Column: 10},
				},
			},
			{
				Code: "describe(\"foo\", () => {\n  it(`catches backticks with the same title`, () => {});\n  it(`catches backticks with the same title`, () => {});\n});\n",
				Errors: []rule_tester.InvalidTestCaseError{
					{MessageId: "multipleTestTitle", Line: 3, Column: 6},
				},
			},
			{
				Code: `context('foo', () => {
  describe('foe', () => {});
});
describe('foo', () => {});
`,
				Settings: map[string]interface{}{
					"jest": map[string]interface{}{
						"globalAliases": map[string]interface{}{
							"describe": []interface{}{"context"},
						},
					},
				},
				Errors: []rule_tester.InvalidTestCaseError{
					{MessageId: "multipleDescribeTitle", Line: 4, Column: 10},
				},
			},
			{
				Code: `spec('works', () => {});
it('works', () => {});
`,
				Settings: map[string]interface{}{
					"jest": map[string]interface{}{
						"globalAliases": map[string]interface{}{
							"it": []interface{}{"spec"},
						},
					},
				},
				Errors: []rule_tester.InvalidTestCaseError{
					{MessageId: "multipleTestTitle", Line: 2, Column: 4},
				},
			},
		},
	)
}

File: packages/rslint-test-tools/tests/eslint-plugin-jest/rules/no-identical-title.test.ts
import { RuleTester } from '../rule-tester';

const ruleTester = new RuleTester();

ruleTester.run('no-identical-title', {} as never, {
  valid: [
    { code: 'it(); it();' },
    { code: 'describe(); describe();' },
    { code: 'describe("foo", () => {}); it("foo", () => {});' },
    {
      code: `
      describe("foo", () => {
        it("works", () => {});
      });
    `,
    },
    {
      code: `
      it('one', () => {});
      it('two', () => {});
    `,
    },
    {
      code: `
      describe('foo', () => {});
      describe('foe', () => {});
    `,
    },
    {
      code: `
      it(\`one\`, () => {});
      it(\`two\`, () => {});
    `,
    },
    {
      code: `
      describe(\`foo\`, () => {});
      describe(\`foe\`, () => {});
    `,
    },
    {
      code: `
      describe('foo', () => {
        test('this', () => {});
        test('that', () => {});
      });
    `,
    },
    {
      code: `
      test.concurrent('this', () => {});
      test.concurrent('that', () => {});
    `,
    },
    {
      code: `
      test.concurrent('this', () => {});
      test.only.concurrent('that', () => {});
    `,
    },
    {
      code: `
      test.only.concurrent('this', () => {});
      test.concurrent('that', () => {});
    `,
    },
    {
      code: `
      test.only.concurrent('this', () => {});
      test.only.concurrent('that', () => {});
    `,
    },
    {
      code: `
      test.only('this', () => {});
      test.only('that', () => {});
    `,
    },
    {
      code: `
      describe('foo', () => {
        it('works', () => {});

        describe('foe', () => {
          it('works', () => {});
        });
      });
    `,
    },
    {
      code: `
      describe('foo', () => {
        describe('foe', () => {
          it('works', () => {});
        });

        it('works', () => {});
      });
    `,
    },
    { code: "describe('foo', () => describe('foe', () => {}));" },
    {
      code: `
      describe('foo', () => {
        describe('foe', () => {});
      });

      describe('foe', () => {});
    `,
    },
    { code: 'test("number" + n, function() {});' },
    {
      code: 'test("number" + n, function() {}); test("number" + n, function() {});',
    },
    { code: 'it(`${n}`, function() {});' },
    { code: 'it(`${n}`, function() {}); it(`${n}`, function() {});' },
    {
      code: `
      describe('a class named ' + myClass.name, () => {
        describe('#myMethod', () => {});
      });

      describe('something else', () => {});
    `,
    },
    {
      code: `
      describe('my class', () => {
        describe('#myMethod', () => {});
        describe('a class named ' + myClass.name, () => {});
      });
    `,
    },
    {
      code: `
      describe("foo", () => {
        it(\`ignores $\{someVar} with the same title\`, () => {});
        it(\`ignores $\{someVar} with the same title\`, () => {});
      });
    `.replace(/\\\{/u, '{'),
    },
    {
      code: `
      const test = { content: () => "foo" };
      test.content(\`something that is not from jest\`, () => {});
      test.content(\`something that is not from jest\`, () => {});
    `,
    },
    {
      code: `
      const describe = { content: () => "foo" };
      describe.content(\`something that is not from jest\`, () => {});
      describe.content(\`something that is not from jest\`, () => {});
    `,
    },
    {
      code: `
      describe.each\`
        description
        $\{'b'}
      \`('$description', () => {});

      describe.each\`
        description
        $\{'a'}
      \`('$description', () => {});
    `,
    },
    {
      code: `
      describe('top level', () => {
        describe.each\`\`('nested each', () => {
          describe.each\`\`('nested nested each', () => {});
        });

        describe('nested', () => {});
      });
    `,
    },
    {
      code: `
      describe.each\`\`('my title', value => {});
      describe.each\`\`('my title', value => {});
      describe.each([])('my title', value => {});
      describe.each([])('my title', value => {});
    `,
    },
    {
      code: `
      describe.each([])('when the value is %s', value => {});
      describe.each([])('when the value is %s', value => {});
    `,
    },
  ],
  invalid: [
    {
      code: `
        describe('foo', () => {
          it('works', () => {});
          it('works', () => {});
        });
      `,
      errors: [{ messageId: 'multipleTestTitle', column: 6, line: 3 }],
    },
    {
      code: `
        it('works', () => {});
        it('works', () => {});
      `,
      errors: [{ messageId: 'multipleTestTitle', column: 4, line: 2 }],
    },
    {
      code: `
        test.only('this', () => {});
        test('this', () => {});
      `,
      errors: [{ messageId: 'multipleTestTitle', column: 6, line: 2 }],
    },
    {
      code: `
        xtest('this', () => {});
        test('this', () => {});
      `,
      errors: [{ messageId: 'multipleTestTitle', column: 6, line: 2 }],
    },
    {
      code: `
        test.only('this', () => {});
        test.only('this', () => {});
      `,
      errors: [{ messageId: 'multipleTestTitle', column: 11, line: 2 }],
    },
    {
      code: `
        test.concurrent('this', () => {});
        test.concurrent('this', () => {});
      `,
      errors: [{ messageId: 'multipleTestTitle', column: 17, line: 2 }],
    },
    {
      code: `
        test.only('this', () => {});
        test.concurrent('this', () => {});
      `,
      errors: [{ messageId: 'multipleTestTitle', column: 17, line: 2 }],
    },
    {
      code: `
        describe('foo', () => {});
        describe('foo', () => {});
      `,
      errors: [{ messageId: 'multipleDescribeTitle', column: 10, line: 2 }],
    },
    {
      code: `
        describe('foo', () => {});
        xdescribe('foo', () => {});
      `,
      errors: [{ messageId: 'multipleDescribeTitle', column: 11, line: 2 }],
    },
    {
      code: `
        fdescribe('foo', () => {});
        describe('foo', () => {});
      `,
      errors: [{ messageId: 'multipleDescribeTitle', column: 10, line: 2 }],
    },
    {
      code: `
        describe('foo', () => {
          describe('foe', () => {});
        });
        describe('foo', () => {});
      `,
      errors: [{ messageId: 'multipleDescribeTitle', column: 10, line: 4 }],
    },
    {
      code: `
        describe("foo", () => {
          it(\`catches backticks with the same title\`, () => {});
          it(\`catches backticks with the same title\`, () => {});
        });
      `,
      errors: [{ messageId: 'multipleTestTitle', column: 6, line: 3 }],
    },
    // TODO: Add this test case when we support global aliases
    // {
    //   code: `
    //     context('foo', () => {
    //       describe('foe', () => {});
    //     });
    //     describe('foo', () => {});
    //   `,
    //   errors: [{ messageId: 'multipleDescribeTitle', column: 10, line: 4 }],
    //   settings: { jest: { globalAliases: { describe: ['context'] } } },
    // },
  ],
});
